opensquirrel/squirrel_ir.py

```diff
@@ -173,8 +170,10 @@ def __init__(self, matrix: np.ndarray, operands: List[Qubit], generator=None, ar
 
     def __eq__(self, other):
         # TODO: Determine whether we shall allow for a global phase difference here.
-        if not isinstance(other, MatrixGate):
+        if not isinstance(other, ControlledGate | MatrixGate):
```

Comment:
Should a `ControlledGate` inherit from `MatrixGate`? Like, all `ControlledGate`s are `MatrixGate`s. Just asking!

Reply:
OK, thanks for the explanation Pablo. So we have here three disjoint set of gates then.

I understand this could be an implementation detail turn, but it appears to me that a ControlledGate would fit very well the role of _**decorator**_ of a gate (a modifier). So, the same as you have donuts and chocolate covered donuts, you would have single-qubit unitaries, e.g. Z, and controlled single-qubit unitaries, in this example controlled Zs.

Reply:
I think you are right about the decorator, but it's already pretty much the case
This decorator still needs to inherit from `Gate` since it is a gate
And it needs to store a control qubit and a target gate, which is already done

Except if i'm missing something I think `ControlledGate` is already a "decorator" although we could make it more explicit by creating a `GateDecorator` interface from which `ControlledGate` would inherit

Then we can e.g. add the `PowerGate` decorator...